src/main/java/edu/kit/ipd/crowdcontrol/objectservice/moneytransfer/MoneyTransferManager.java

```diff
@@ -0,0 +1,172 @@
+        this.mailHandler = mailHandler;
+        this.workerOperations = workerOperations;
+        this.workerBalanceOperations = workerBalanceOperations;
+        this.minGiftCodesCount = 10;
+        this.payOffThreshold = 0;
+        this.notificationMailAddress = notificationMailAddress;
+        this.notificationText = new StringBuilder();
+    }
+
+    /**
+     * Logs a new money transfer and saves it.
+     *
+     * @param workerID the id of the worker, who gets the money
+     * @param amount   the amount of money in ct
+     */
+    public void logMoneyTransfer(int workerID, int amount, int expID) {
+        workerBalanceOperations.addCredit(workerID, amount, expID);
+    }
+
+    /**
+     * Pays all workers depending on their logged money transfers.
+     */
+    public void payOff() {
+        fetchNewGiftCodes();
+
+        Result<WorkerRecord> workers = workerOperations.getWorkerWithCreditBalanceGreaterOrEqual(payOffThreshold);
+        Iterator<WorkerRecord> workerIt = workers.iterator();
+        List<GiftCodeRecord> giftCodes = workerBalanceOperations.getUnusedGiftCodes();
+
+        while (workerIt.hasNext()) {
+            WorkerRecord worker = workerIt.next();
+            List<GiftCodeRecord> payedCodesForWorker = chooseGiftCodes(worker, giftCodes);
+
+            giftCodes = workerBalanceOperations.getUnusedGiftCodes();
+            payWorker(worker, payedCodesForWorker);
+        }
+        if (giftCodes.size() < minGiftCodesCount) {
+            notificationText.append("There are less than ").append(minGiftCodesCount).append(" giftcodes in the database. It is recommended to add more.").append(System.getProperty("line.separator"));
+        }
+
+        sendNotification();
+    }
+
+    private void fetchNewGiftCodes() {
+        Message[] messages = new Message[0];
+        try {
+            messages = mailHandler.fetchUnseen("inbox");
+        } catch (MessagingException e) {
+            e.printStackTrace();
+        }
+
+        for (Message message : messages) {
+            try {
+                GiftCodeRecord rec = MailParser.parseAmazonGiftCode(message);
+                workerBalanceOperations.addGiftCode(rec.getCode(), rec.getAmount());
+            } catch (MessagingException | IOException e) {
+                e.printStackTrace();
+            } catch (AmazonMailFormatChangedException e) {
+                try {
+                    mailHandler.markAsUnseen(message);
+                } catch (MessagingException f) {
+                    f.printStackTrace();
+                }
+                notificationText.append("It seems, that amazon changed the format of the giftcode E-Mails. You need to adjust the parser to import giftcodes in future.");
+            }
+        }
+    }
+
+    private List<GiftCodeRecord> chooseGiftCodes(WorkerRecord worker, List<GiftCodeRecord> giftCodes) {
+        List<GiftCodeRecord> payedCodes = new LinkedList<>();
```

Comment:
yeah, they are only useful if you have extremely large lists and you are only adding/removing on head and tail. But then you should most likely use a queue.